### PR TITLE
Dataaccessapi

### PR DESCRIPTION
**Pull request by dan-sh**
The Data Access API Upgrade 

Data Access functionality to allow simple data retrieval across TorQ processes, see docs/dataaccess.md for full specification.

(The link to the blog in the docs will be updated on release.)

### CODE REVIEW

File: code/common/checkinputs.q [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+    if[in[`sqlquery;key dict];:isstring[dict;`sqlquery]];
+    dict:checkdictionary dict;
+    dict:checkinvalidcombinations dict;
+    dict:checkrepeatparams dict;
+    dict:checkeachparam[dict;1b];
+    dict:checkeachparam[dict;0b];
+    dict[`checksperformed]:1b;
+    :dict;
+  };
+
```
Comment (drgdavies) [inline code/common/checkinputs.q:19]: Why are comments on some functions and not others?

**Comment by drgdavies** [comment]
There doesn't seem to be much logging through the code base for the size of it.  If an issue occurs in the code how do you know how far it has progressed or where it is?

File: code/common/checkinputs.q [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+// check tablename parameter is of type symbol
+checktable:{[dict;parameter]:checktype[-11h;dict;parameter];};
+
+// check that endtime is of temporal type and that it is greater than or equal to starttime
+checkendtime:{[dict;parameter]
+    dict:checktimetype[dict;parameter];
+    :checktimeorder dict};
+
+// check that inputted value is of valid type: -12 -14 -15h
+checktimetype:{[dict;parameter]:checktype[-12 -14 -15h;dict;parameter];};
```
Comment (drgdavies) [inline code/common/checkinputs.q:75]: Might be better to have a global parameter of time types so that if KX decided to add/remove its easy to find and add to the list.

File: tests/dataaccess/queryorder/settings.q [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+//- extract the input dictionary from {testname}.csv
+//- extract the respone from .queryorder.orderquery
+//- compare output with expected one
+
+
+getinputparams:{[test]exec parameter!get each parametervalue from .checkinputs.readcsv[` sv inputpath,`$string[test],".csv";"s*"]};
+
+getoutputparams:{[test]T:exec parameter!get each parametervalue from .checkinputs.readcsv[` sv outputpath,`$string[test],".csv";"i*"];:(T[til 4])};
+
+
```
Comment (drgdavies) [inline tests/dataaccess/queryorder/settings.q:15]: Excess white space throughout file

File: tests/dataaccess/queryorder/queryorder.csv [hunk 4]
```diff
@@ -0,0 +1,24 @@
+action,ms,bytes,lang,code,repeat,minver,comment
```
Comment (drgdavies) [inline tests/dataaccess/queryorder/queryorder.csv:1]: Is this all the tests we have for this?  Does it really cover all scenarios?

File: tests/dataaccess/queryorder/hdb/dbmaint.q [hunk 5]
```diff
@@ -0,0 +1,149 @@
+/ kdb+ partitioned database maintenance
+\d .os
+WIN:.z.o in`w32`w64
+pth:{p:$[10h=type x;x;string x];if[WIN;p[where"/"=p]:"\\"];(":"=first p)_ p}
+cpy:{system$[WIN;"copy /v /z ";"cp "],pth[x]," ",pth y}
```
Comment (drgdavies) [inline tests/dataaccess/queryorder/hdb/dbmaint.q:5]: Why is this file needed here?  Surely its code duplication?

File: tests/dataaccess/mockdata.q [hunk 6]
```diff
@@ -0,0 +1,61 @@
+params:([proctype:`hdb`rdb]
+  func:`generatehdb`generaterdb;
+  partitiontype:`date`date;
+  hdbname:`hdb`;
+  n:5 5;
+  tablename:`xdaily`xdaily;
+  nrecord:10 10
+ );
```
Comment (drgdavies) [inline tests/dataaccess/mockdata.q:8]: No comments or purpose of this file?

File: tests/dataaccess/common/testdata/checkinputerrors.csv [hunk 7]
```diff
@@ -0,0 +1,41 @@
+test,error
+nondictinput,Input must be a dictionary
```
Comment (drgdavies) [inline tests/dataaccess/common/testdata/checkinputerrors.csv:2]: Why are these specified like this?   this says checkinputerrors?

File: tests/dataaccess/common/settings.q [hunk 8]
```diff
@@ -0,0 +1,12 @@
+testpath:hsym`$getenv[`KDBTESTS],"/dataaccess/common";
+processcsv:` sv testpath,`config`process.csv;
+
+//- code to pass in a test name
+//- extract the input parameter from {testname}.csv
+//- extract the expected error from checkinputerrors.csv
+//- compare error with expected error
+checkreturnederror:{[test]errors[test;`error]~@[.checkinputs.checkinputs;gettestparams test;::]};
```
Comment (drgdavies) [inline tests/dataaccess/common/settings.q:8]: Is this not common unit test code?

File: code/gateway/dataaccess.q [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+timebarmap:`nanosecond`timespan`microsecond`second`minute`hour`day!1 1 1000 1000000000 60000000000 3600000000000 86400000000000;
+
+// function to convert sorting
+go:{if[`asc=x[0];:(xasc;x[1])];:(xdesc;x[1])};
+
+// Full generality dataaccess function in the gateway
+getdata:{[o]
+    // Input checking in the gateway
+    reqno:.requests.initlogger[o];
```
Comment (drgdavies) [inline code/gateway/dataaccess.q:13]: Are we recording requests?

File: code/gateway/dataaccess.q [hunk 10]
```diff
@@ -0,0 +1,107 @@
+\d .dataaccess
+
+forceservers:0b
+
+timebarmap:`nanosecond`timespan`microsecond`second`minute`hour`day!1 1 1000 1000000000 60000000000 3600000000000 86400000000000;
```
Comment (drgdavies) [inline code/gateway/dataaccess.q:5]: This is duplicated elsewhere, surely doesn't need defined multiple times?

File: code/gateway/daqrest.q [hunk 11]
```diff
@@ -0,0 +1,44 @@
+.aqrest.execute:{[req;props] @[value;req;{(neg .z.w)(.gw.formatresponse[0b;0b;"error: ",x])}]};
+
+\d .dataaccess
+
+// .gw.formatresponse:{[status;sync;result]$[not[status]and sync;'result;result]}};
```
Comment (drgdavies) [inline code/gateway/daqrest.q:5]: Is this a comment or a commented out line of code?

File: code/dataaccess/extractqueryparams.q [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+  if[typ~15h;:timebucket xbar`timespan$x];
+  if[typ in 16 17 18 19h;:timebucket xbar`timespan$x];
+  '`$"timebar type error"; //- type checks in checkinputs functions should stop it reaching here
+ };
+
+// extract where filter parameters from input dictionary
+// the filters parameter is a dictionary of the form:
+// `sym`price`size!(enlist(=;`AAPL);((within;80 100);(not in;81 83 85));enlist(>;50))
+// this is translated into a kdb parse tree for use in the where clause:
+// ((=;`sym;,`AAPL);(within;`price;80 100);(in[~:];`price;81 83 85);(>;`size;50))
```
Comment (drgdavies) [inline code/dataaccess/extractqueryparams.q:114]: What is the determination on what gets commented function-wise?

File: code/dataaccess/extractqueryparams.q [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+  timebar:`size`bucket`timecol!inputparams`timebar;
+  // Convert the timebucket to it's corresponding integer value 
+  timebucket:exec size * .dataaccess.timebarmap bucket from timebar;
+  // Return as a kdb+ native function
+  :@[queryparams;`timebar;:;timebar[1#`timecol]!enlist(xbarfunc;timebucket;timebar[`timecol])];
+  };
+
+xbarfunc:{[timebucket;x]
+  typ:type x;
+  if[typ~12h;:timebucket xbar x];
```
Comment (drgdavies) [inline code/dataaccess/extractqueryparams.q:103]: Use checktype?

File: code/dataaccess/customfuncs.q [hunk 14]
```diff
@@ -0,0 +1,44 @@
+//- Script to load in custom functionality:
+//- Useful for splitting queries 
+
+\d .dacustomfuncs
+
+//- (i) rollover
+//- Function to determine which partitions the getdata function should query
+//- e.g If the box is based in Paris (GMT+01:00) and rollover is at midnight London time then tzone:-01:00 
+//- e.g If the box is UTC based and rollover is at 10pm UTC then rover: 22:00
+
```
Comment (drgdavies) [inline code/dataaccess/customfuncs.q:10]: Whitespace

File: code/dataaccess/customfuncs.q [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    // convert from data timezone -> utc
+    adjdata:exec adjustment from .tz.t asof `timezoneID`gmtDateTime!(A[`datatimezone];hdbtime+adjroll);
+    querytimeUTC:`time$hdbtime+$[0Nn~adjdata;00:00;adjdata];
+    $[querytimeUTC >= rolltimeUTC;:A[`partitionfield]$hdbtime;:offsetbyone[hdbtime;A[`partitionfield]]];              
+    };
+
+//- (ii) getpartitionrange
+//- offset times for non-primary time columns
+// example @[`date$(starttime;endtime);1;+;not `time~`time]
+
```
Comment (drgdavies) [inline code/dataaccess/customfuncs.q:29]: Whitespace

**Comment by dan-sh** [comment]
> There doesn't seem to be much logging through the code base for the size of it. If an issue occurs in the code how do you know how far it has progressed or where it is?

Logging is done at the [getdata level](https://github.com/AquaQAnalytics/TorQ/blob/dataaccessapi/code/dataaccess/getdata.q#L10)

`inputparams:@[.dataaccess.checkinputs;inputparams;.requests.error[requestnumber;]];`

The theory is if an error occurs it will occur in the checkinputs stage. This error will then propagate into the `.dataaccess.stats` with the error- would you say this is sufficient?

File: code/common/checkinputs.q [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+// check that timebar parameter has three elements of respective types: numeric, symbol,
+// symbol.
+checktimebar:{[dict;parameter]
+    input:dict parameter;
+    if[not(3=count input)&0h~type input;
+        '`$"Timebar parameter must be a list of length 3"];
+    input:`size`bucket`timecol!input;
+    if[not any -6 -7h~\:type input`size;
+        '`$"First argument of timebar must be either -6h or -7h"];
+    if[not -11h~type input`bucket;
```
[the remark this answers is not in the document]
Comment (robertcoen) [reply code/common/checkinputs.q:120]: checktype takes in three arguments; valid types, the query as a dictionary and the parameter to be checked. An error message is generated from these arguments. The timebar parameter is a mixed list of one number and two symbols and checktimebar checks each element in timebar is of the correct type. checktype can only check if timebar is type 0h

File: code/dataaccess/extractqueryparams.q (hunk 13 above)
Comment (dan-sh) [reply code/dataaccess/extractqueryparams.q:103]: We need the boolean

File: code/gateway/daqrest.q (hunk 11 above)
Comment (dan-sh) [reply code/gateway/daqrest.q:5]: It's the default definition of the function would you rather it deleted?

File: code/gateway/dataaccess.q (hunk 9 above)
Comment (dan-sh) [reply code/gateway/dataaccess.q:13]: Yeah the reqno is essentially a ticket number for which entry in the logging table the query relates. If logging is turned off the 'ticketnumber' is set to -1 and the requests aren't logged.

File: code/common/checkinputs.q [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+    if[not checkparamnames dict;'`$.checkinputs.formatstring["Invalid parameter present:{}. For a list of valid parameters, use the command .checkinputs.getvalidparams[]";key[dict]except .checkinputs.getvalidparams[]]];
+    :dict;
+  };
+
+isdictionary:{[dict]$[99h~type dict;:dict;'`$"Input must be a dictionary"]};
+checkkeytype:{[dict]11h~type key dict};
+checkrequiredparams:{[dict]all .checkinputs.getrequiredparams[]in key dict};
+getrequiredparams:{[]exec parameter from .checkinputs.checkinputsconfig where required}
+checkparamnames:{[dict]all key[dict]in .checkinputs.getvalidparams[]};
+getvalidparams:{[]exec parameter from .checkinputs.checkinputsconfig};
```
Comment (jonathonmcmurray) [inline code/common/checkinputs.q:32]: Are these functions used frequently/outside of the above `checkinputs` & `checkdictionary` functions?

They're all quite short, this is probably debatable but IMO they're short enough that it's not worth turning these into separate functions unless they're being used a lot.

File: code/common/utils.q [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+  tlist:exec tablename from .checkinputs.tablepropertiesconfig;
+  partfield:$[()~key`.Q.pf;`;.Q.pf];
+  metainfo:1!/:`columns`types`attributes xcol/:`c`t`a#/:0!/:meta each tlist;
+  :1!flip(`tablename`partfield`metas`proctype)!(tlist;partfield;metainfo;.proc.proctype);
+  .lg.o[`getmetainfo;"metainfo successfully loaded"]
+ };
+
+//- misc utils
+getvalidparams:{[]checkinputsconfig`parameter};
+getrequiredparams:{[]exec parameter from checkinputsconfig where required};
```
Comment (jonathonmcmurray) [inline code/common/utils.q:45]: aren't these defined in `code/common/checkinputs.q`? do they need to be defined here as well? (as far as I can see with a quick search they aren't referenced elsewhere in this file)

File: code/common/dataaccess.q [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+
+//- init function takes:
+//- (i) tablepropertiespath - either from a valid -dataaccess path from cmd line, or passed explicitly
+//-                           otherwise a user can call "init[`:/path/to/tableproperties.csv]"
+//- init will:
+//-   - load data access code from code/common dataaccess
+//-   - validate table properties config file exists + load it 
+//-   - load config for checking input parameters
+//-   - write meta info for tables in current process to .dataaccess.metainfo
+
```
[the remark this answers is not in the document]
Comment (dan-sh) [reply code/common/dataaccess.q]: Done

File: code/common/dataaccess.q [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+\d .
+
+if[.proc.proctype in `rdb`hdb`gateway;
+  // set table properties path
+  .dataaccess.settablepropertiespath[];
+  // re-initialize on new connections 
+  if[.dataaccess.validtablepropertiespath[];.servers.connectcustom:.dataaccess.connectcustom];
+  ];
+
+\d .schema
```
[the remark this answers is not in the document]
Comment (dan-sh) [reply code/common/dataaccess.q:56]: Done

File: code/common/utils.q [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+  .lg.o[`readtableproperties;"loading table properties"];
+  table:`tablename`proctype xkey readcsv[tablepropertiepath;"ssssstsss"];                                                            //read in table from file
+  alltable:?[table;enlist(in;`proctype;enlist`all`);0b;()];                                                                          //find any instance of the use "all" or blank for proctype
+  table:table,![alltable;();0b;(enlist`proctype)!enlist(enlist `hdb)],![alltable;();0b;(enlist`proctype)!enlist(enlist `rdb)];       //join rdb and hdb entries for any "all" or blank entries
+  table:![table;enlist(in;`proctype;enlist`all`);0b;`symbol$()];                                                                     //remove "all" or blank entries from table
+  table:?[table;$[.proc.proctype=`gateway;();enlist(=;`proctype;`.proc.proctype)];0b;()];
+  table:update pfield^partitionfield,.eodtime.datatimezone ^ datatimezone, .eodtime.rolltimeoffset ^ rolltimeoffset,.eodtime.rolltimezone^rolltimezone from table;
+  .lg.o[`readtableproperties;"Table properties successfully loaded"];
+  :table;
+      };
```
[the remark this answers is not in the document]
Comment (dan-sh) [reply code/common/utils.q]: Done

File: code/dataaccess/checkinputs.q [hunk 22]
```diff
@@ -0,0 +1,150 @@
+\d .dataaccess
+
+// checkinputs is the main function called when running a query - it checks:
+//   (i) input format
+//  (ii) whether any parameter pairs clash
+// (iii) parameter specific checks
+// The input dictionary accumulates some additional table information/inferred info
+checkinputs:{[dict]
```
[the remark this answers is not in the document]
Comment (dan-sh) [reply code/dataaccess/checkinputs.q:8]: Further logging for the three major functions has been added- this happens at the getdata level.

File: tests/dataaccess/queryorder/queryorder.csv (hunk 4 above)
Comment (dan-sh) [reply tests/dataaccess/queryorder/queryorder.csv:1]: The piece of coded we are testing, [Queryorder](https://github.com/AquaQAnalytics/TorQ/blob/dataaccessapi/code/dataaccess/queryorder.q), is a really light piece of code- it only reorders the functional select.

File: tests/dataaccess/common/settings.q (hunk 8 above)
Comment (dan-sh) [reply tests/dataaccess/common/settings.q:8]: No our errors are only caught in one place- checkinputs, hence we only need to trap errors in this script. It is not part of the [greater framework](https://github.com/AquaQAnalytics/TorQ/blob/dataaccessapi/tests/helperfunctions.q) or [K4](https://code.kx.com/q/kb/unit-tests/)

File: code/common/utils.q (hunk 18 above)
Comment (robertcoen) [reply code/common/utils.q:44]: There was code lifted from `checkinputs.q` and moved to `utils.q` . I'll remove the duplicate code.

File: code/gateway/daqrest.q (hunk 11 above)
Comment (drgdavies) [reply code/gateway/daqrest.q:5]: If its needed keep it but make it clear its a comment versus a commented line of code

File: tests/dataaccess/common/testdata/checkinputerrors.csv (hunk 7 above)
Comment (robertcoen) [reply tests/dataaccess/common/testdata/checkinputerrors.csv:2]: Checkinputerrors.csv loads in the dictionary of expected error messages to check against query error messages. These two messages are checked against each other to ensure that they match.

File: code/common/checkinputs.q (hunk 16 above)
[the remark this answers is not in the document]
Comment (drgdavies) [reply code/common/checkinputs.q:120]: Ok - did you manage to confirm if it could be tested using checktype instead?

[the remark this answers is not in the document]
Comment (robertcoen) [reply code/common/checkinputs.q:120]: We can if we define a new dictionary using the timebar parameter, the only catch is that it would display the error message as first_argument_of_timebar input type incorrect - valid type(s):-6h -7h - input type:-11h

File: code/gateway/dataaccess.q (hunk 10 above)
Comment (dan-sh) [reply code/gateway/dataaccess.q]: Done

File: code/common/utils.q [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+
+readcheckinputs:{[checkinputspath] spliltcolumns[readcsv[checkinputspath;"sbs*"];`invalidpairs;`]};
+
+readcsv:{[path;types]
```
[the remark this answers is not in the document]
Comment (dan-sh) [reply code/common/utils.q:18]: Suprisingly not

File: code/common/checkinputs.q [hunk 24]
```diff
@@ -0,0 +1,257 @@
...
+    .checkinputs.checktype[-11h;input;`$"timebar third arguement"];
+    :dict;
+  };
+
+// check that filters parameter is of type dictionary, has symbol keys, the values are 
+// in (where function;value(s)) pairs and (not)within filtering functions have two values
+// associated with it.
+checkfilters:{[dict;parameter]
+    dict:checktype[99h;dict;parameter];
+    (dict parameter):@[(dict parameter);where {not all 0h=type each x}each (dict parameter);enlist];
```
Comment (tsmyth-aquaq) [inline code/common/checkinputs.q:121]: Convention is for square brackets:
```suggestion
    dict[parameter]:@[(dict parameter);where {not all 0h=type each x}each (dict parameter);enlist];
```
Also wondering if you can do something like this
```
dict:@[dict;parameter;{$[not all 0h=type each x;enlist;]x}];
```

File: code/common/checkinputs.q [hunk 25]
```diff
@@ -0,0 +1,257 @@
...
+        '`$errmess2];
+    if[(((enlist"=")~string first pair)|(enlist"~")~string first pair)&1<>count last pair;
+        '`$errmess2];
+    if[("~="~string first pair)&1<>count last pair;
+        '`$errmess];};
+
+// check that ordering parameter contains only symbols and is paired in the format
+// (direction;column).
+checkordering:{[dict;parameter]
+    if[11h=type (dict parameter);(dict parameter):enlist (dict parameter)];
```
Comment (tsmyth-aquaq) [inline code/common/checkinputs.q:159]: Convention is for square brackets + too many brackets anyway
```suggestion
    if[11h=type dict parameter;dict[parameter]:enlist dict parameter];
```

File: code/common/checkinputs.q [hunk 26]
```diff
@@ -0,0 +1,257 @@
...
+    input:dict parameter;
+    if[11h<>type raze input;
+        '`$.schema.errors[`checkorderingpair;`errormessage],.schema.examples[`ordering1;`example]];
+    if[0<>count except[count each input;2];
+        '`$.schema.errors[`checkorderingarrangment;`errormessage],.schema.examples[`ordering1;`example]];
+    if[0<>count except[first each input;`asc`desc];
+        '`$.schema.errors[`checkorderingdirection;`errormessage],.schema.examples[`ordering1;`example]];
+    $[in[`grouping;key dict];grouping:(dict`grouping),();grouping:()];
+    $[in[`timebar;key dict];timebar:(dict`timebar);timebar:()];
```
Comment (tsmyth-aquaq) [inline code/common/checkinputs.q:168]: Can reduce duplication and I see no reason not to use infix notation.
```suggestion
    grouping:$[`grouping in key dict;(),dict`grouping;()];
    timebar:$[`timebar in key dict;dict`timebar;()];
```

File: code/common/checkinputs.q [hunk 27]
```diff
@@ -0,0 +1,257 @@
...
+    if[in[`aggregations;key dict];
```
Comment (tsmyth-aquaq) [inline code/common/checkinputs.q:169]: Use infix?
```suggestion
    if[`aggregations in key dict;
```

File: code/common/checkinputs.q [hunk 28]
```diff
@@ -0,0 +1,257 @@
...
+        aggs:dict`aggregations;
+        aggs:flip(key[aggs]where count each get aggs;raze aggs);
+        names:{
+        if[1=count x[1];
+            :`$(string x[0]),@[string x[1];0;upper]];
+        if[2=count x[1];
+            :`$(string x[0]),(@[string first x[1];0;upper]),@[string last x[1];0;upper]]
+        }'[aggs];
```
Comment (tsmyth-aquaq) [inline code/common/checkinputs.q:177]: Indentation needs fixed and I think this works out the same. If the count is always 1 or 2 can drop the if condition.
```suggestion
        names:{
            if[count[x 1]in 1 2;
                :`$raze string[x 0],.[string (),x 1;(::;0);upper]
        }'[aggs];
```

File: code/common/checkinputs.q [hunk 29]
```diff
@@ -0,0 +1,257 @@
...
+        if[any raze {1<sum y=x}[last each aggs]'[last each input];
+            '`$.schema.errors[`orderingvague;`errormessage],.schema.examples[`ordering2;`example]];
+        if[any not in[last each input;names,grouping,timebar[2],last each aggs];
+            '`$.schema.errors[`orderingnocolumn;`errormessage]]];
+    if[in[`columns;key dict];
+        columns:(dict`columns),();
+        if[not (enlist `)~columns;
+            if[any not in[last each input;columns];
+                badorder:sv[",";string (last each input) where not in[last each input;columns]]; 
+                '`$.checkinputs.formatstring[.schema.errors[`badorder;`errormessage];`$badorder]]]];
```
Comment (tsmyth-aquaq) [inline code/common/checkinputs.q:187]: Can reduce some bulk and use infix notation:
```suggestion
    if[in[`columns;key dict];
        if[not enlist[`]~columns:(),dict`columns;
            if[any not l:last'[input]in columns;
                badorder:","sv string last'[input]where l; 
                '`$.checkinputs.formatstring[.schema.errors[`badorder;`errormessage];`$badorder]]]];
```

File: code/common/checkinputs.q [hunk 30]
```diff
@@ -0,0 +1,257 @@
...
+    :dict;};
+    
+// check that the instrumentcol parameter is of type symbol
+checkinstrumentcolumn:{[dict;parameter]:checktype[-11h;dict;parameter];};
+
+checkrenamecolumn:{[dict;parameter]
+    dict:checktype[99 -11 11h;dict;parameter];
+    input:dict parameter;
+    if[(type input) in (-11h;11h);:dict];
```
Comment (tsmyth-aquaq) [inline code/common/checkinputs.q:196]:
```suggestion
    if[type[input]in -11 11h;:dict];
```

File: code/common/utils.q [hunk 31]
```diff
@@ -0,0 +1,82 @@
...
+  if[not property in key inputparams subdict;'`$"gettableproperty:invalid property"];
+  :inputparams[subdict;property];
+ };
+
+gettableproperty:extractfromsubdict[;`tableproperties;];   //- extract from `tableproperties key in inputparams
+
+//- get default time from  tickerplant or table
+getdefaulttime:{[dict]
+  if[(dict`tablename) in key .schema;
+    :first ?[meta `$(".schema.",string (dict`tablename));enlist(=;`t;"p");();`c]];
```
Comment (tsmyth-aquaq) [inline code/common/utils.q:74]: Can use symbols instead of strings here:
```suggestion
    :first ?[meta` sv``schema,dict`tablename;enlist(=;`t;"p");();`c]];
```

**Review by drgdavies — APPROVED** [review]
